--- includes/wf_crm_webform_postprocess.inc.orig	2018-02-28 10:19:13.120269628 -0500
+++ includes/wf_crm_webform_postprocess.inc	2018-02-27 13:16:33.238440338 -0500
@@ -797,7 +797,7 @@
               continue;
             }
             // Reset calculated values when updating an address
-            $params['master_id'] = $params['geo_code_1'] = $params['geo_code_2'] = 'null';
+            $params['master_id'] = $params['geo_code_1'] = $params['geo_code_2'] = null;
           }
           $params['contact_id'] = $cid;
           if (!empty($existing[$i])) {
